src/java/org/apache/cassandra/index/sai/memory/RowMapping.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.cassandra.index.sai.utils.PrimaryKeys;
+import org.apache.cassandra.io.compress.BufferType;
+import org.apache.cassandra.utils.AbstractGuavaIterator;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+/**
+ * In memory representation of {@link PrimaryKey} to row ID mappings which only contains
+ * {@link Row} regardless it's live or deleted. ({@link RangeTombstoneMarker} is not included.)
+ *
+ * For JBOD, we can make use of sstable min/max partition key to filter irrelevant {@link MemtableIndex} subranges.
+ * For Tiered Storage, in most cases, it flushes to tiered 0.
+ */
+public class RowMapping
+{
+    public static final RowMapping DUMMY = new RowMapping()
+    {
+        @Override
+        public Iterator<Pair<ByteComparable, LongArrayList>> merge(MemtableIndex index) { return Collections.emptyIterator(); }
+
+        @Override
+        public void complete() {}
+
+        @Override
+        public void add(PrimaryKey key, long sstableRowId) {}
+    };
+
+    private final InMemoryTrie<Long> rowMapping = new InMemoryTrie<>(BufferType.OFF_HEAP);
+
+    private volatile boolean complete = false;
+
+    public PrimaryKey minKey;
+    public PrimaryKey maxKey;
+
+    public long maxSegmentRowId = -1;
+
+    private RowMapping()
+    {}
+
+    /**
+     * Create row mapping for FLUSH operation only.
+     */
+    public static RowMapping create(OperationType opType)
+    {
+        if (opType == OperationType.FLUSH)
+            return new RowMapping();
+        return DUMMY;
+    }
+
+    /**
+     * Merge IndexMemtable(index term to PrimaryKeys mappings) with row mapping of a sstable
+     * (PrimaryKey to RowId mappings).
+     *
+     * @param index a Memtable-attached column index
+     *
+     * @return iterator of index term to postings mapping exists in the sstable
+     */
+    public Iterator<Pair<ByteComparable, LongArrayList>> merge(MemtableIndex index)
+    {
+        assert complete : "RowMapping is not built.";
+
+        Iterator<Pair<ByteComparable, PrimaryKeys>> iterator = index.iterator();
+        return new AbstractGuavaIterator<Pair<ByteComparable, LongArrayList>>()
+        {
+            @Override
+            protected Pair<ByteComparable, LongArrayList> computeNext()
+            {
+                while (iterator.hasNext())
+                {
+                    Pair<ByteComparable, PrimaryKeys> pair = iterator.next();
+
+                    LongArrayList postings = null;
+                    Iterator<PrimaryKey> primaryKeys = pair.right.iterator();
+
+                    while (primaryKeys.hasNext())
+                    {
+                        PrimaryKey primaryKey = primaryKeys.next();
+                        ByteComparable byteComparable = primaryKey::asComparableBytes;
+                        Long segmentRowId = rowMapping.get(byteComparable);
+
+                        if (segmentRowId != null)
```

Comment:
Wouldn't it be *very bad* if there was an indexed primary key without a row ID mapping from the flushing SSTable (i.e. `segmentRowId == null`)? If so, would an assertion w/ a brief message be more appropriate?

Reply:
No, I don't believe that's the case. We don't handle any deletions in the in-memory index so it is entirely possible that the index could contain a primary key that no longer existed in the memtable.

Reply:
Ah, you're right. Thanks.